hermes/probe/journal/journal.go

```diff
@@ -0,0 +1,364 @@
+	"google.golang.org/api/iterator"
+	"google.golang.org/protobuf/testing/protocmp"
+
+	cpmetrics "github.com/google/cloudprober/metrics"
+	pb "github.com/googleinterns/step224-2020/hermes/proto"
+)
+
+const (
+	journalFilePrefix = "Hermes_Journal_" // Prefix of the journal filename.
+	apiLatency        = "hermes_api_latency_seconds"
+)
+
+// journalFile stores the name, checksum and contents of the Journal.
+type journalFile struct {
+	name     string
+	checksum string
+	contents []byte
+}
+
+// checksumContents checksums the contents of the file object and stores
+// the checksum in the checksum field of the object.
+// Returns:
+// 		- err: returned if there is an error when checksumming the contents.
+func (j *journalFile) checksumContents() *probe.ProbeError {
+	h := sha1.New()
+	if _, err := io.Copy(h, j.reader()); err != nil {
+		return probe.NewProbeError(metrics.ProbeFailed, fmt.Errorf("verifyNilFile(): unable to compute checksum of file contents: %w", err))
+	}
+	j.checksum = fmt.Sprintf("%x", h.Sum(nil))
+	return nil
+}
+
+// validateChecksum validates that the checksum matches the checksum
+// in the file name.
+// The checksum in the filename is being validated against the file contents
+// This will return false if the file contents have been updated
+// without the file name being changed accordingly.
+// Returns:
+//	- err: returns err if the checksum does not match or the checksum or filename is blank, else nil.
+func (j *journalFile) validateChecksum() *probe.ProbeError {
+	if j.checksum == "" {
+		return probe.NewProbeError(metrics.ProbeFailed, fmt.Errorf("validateChecksum(): journal checksum cannot be validated as checksum, %q, is blank", j.checksum))
+	}
+	if j.name == "" {
+		return probe.NewProbeError(metrics.ProbeFailed, fmt.Errorf("validateChecksum(): journal checksum cannot be validated as name, %q, is blank", j.name))
+	}
+
+	if j.checksum != j.name[len(journalFilePrefix):] {
+		return probe.NewProbeError(metrics.ProbeFailed, fmt.Errorf("validateChecksum(): checksum %q does not match checksum in name %q", j.checksum, j.name[len(journalFilePrefix):]))
+	}
+
+	return nil
+}
+
+// reader returns an io reader that in turn returns the file contents.
+// Returns:
+// - reader: returns an io reader that in turn returns the file contents.
+func (j *journalFile) reader() io.Reader {
+	return bytes.NewReader(j.contents)
+}
+
+// unmarshalContents unmarshals the contents of the journal file and returns
+// the completed proto struct.
+// Returns:
+// 		- journal: proto struct containing the unmarshalled contents of the journal
+//		- err: error if once occurs during the unmarshalling process.
+func (j *journalFile) unmarshalContents() (*pb.StateJournal, *probe.ProbeError) {
+	journal := &pb.StateJournal{}
+	if err := proto.Unmarshal(j.contents, journal); err != nil {
+		return nil, probe.NewProbeError(metrics.FileReadFailure, fmt.Errorf("verifyNilFile(): unable to unmarshal nil file proto: %w", err))
```

Comment:
Nit: `FileReadFailure` suggests that reading the journal from the storage system failed. I would introduce a different error type for this (unless you already have an error that indicates corruption).

Reply:
There is a `metrics.FileCorrupted` error type, to which I've changed this.